snuba/subscriptions/scheduler.py

```diff
@@ -1,13 +1,259 @@
+        def get_final_mode(transition_mode: TaskBuilderMode) -> TaskBuilderMode:
+            return (
+                TaskBuilderMode.IMMEDIATE
+                if transition_mode == TaskBuilderMode.TRANSITION_IMMEDIATE
+                else TaskBuilderMode.JITTERED
+            )
+
+        def get_start_mode(transition_mode: TaskBuilderMode) -> TaskBuilderMode:
+            return (
+                TaskBuilderMode.IMMEDIATE
+                if transition_mode == TaskBuilderMode.TRANSITION_JITTER
+                else TaskBuilderMode.JITTERED
+            )
+
+        general_mode = TaskBuilderMode(
+            state.get_config("subscription_primary_task_builder", "immediate")
+        )
+
+        if (
+            general_mode == TaskBuilderMode.IMMEDIATE
+            or general_mode == TaskBuilderMode.JITTERED
+        ):
+            return general_mode
+
+        resolution = int(subscription.data.resolution.total_seconds())
+        if resolution > settings.MAX_RESOLUTION_FOR_JITTER:
+            return get_final_mode(general_mode)
+
+        if timestamp % resolution == 0:
+            self.__state[resolution] = get_final_mode(general_mode)
+
+        current_state = self.__state.get(resolution)
+        return (
+            current_state if current_state is not None else get_start_mode(general_mode)
+        )
+
+
+class DelegateTaskBuilder(TaskBuilder[Subscription]):
+    """
+    A delegate capable of switching back and forth between the
+    immediate and jittered task builders according to runtime
+    settings.
+
+    It relies on TaskBuilderModeState to decide which task builder
+    to use. The reason for this is that we cannot simply switch
+    from one mode to another at any point in time. We need to wait
+    for the end of the resolution time interval, or we risk to
+    skip some queries.
+
+    Example: if we transitioned from jittered to immediate at the
+    second 30 of a minute. A query scheduled with a jitter = 40
+    would not be scheduled at all during that minute because,
+    with the immediate task builder, that would be scheduled at
+    second 0.
+    """
+
+    def __init__(self) -> None:
+        self.__immediate_builder = ImmediateTaskBuilder()
+        self.__jittered_builder = JitteredTaskBuilder()
+        self.__rollout_state = TaskBuilderModeState()
+
+    def get_task(
+        self, subscription: Subscription, timestamp: int
+    ) -> Optional[ScheduledTask[Subscription]]:
+        immediate_task = self.__immediate_builder.get_task(subscription, timestamp)
+        jittered_task = self.__jittered_builder.get_task(subscription, timestamp)
+        primary_builder = self.__rollout_state.get_current_mode(subscription, timestamp)
+        if primary_builder == TaskBuilderMode.JITTERED:
+            return jittered_task
+        else:
+            return immediate_task
+
+    def reset_metrics(self) -> Sequence[Tuple[str, int, Tags]]:
+        def add_tag(tags: Tags, builder_type: str) -> Tags:
```

Comment:
Neither immediate or jittered actually returns tags, maybe this could be added at a later date if tags become useful?

Reply:
It is true that neither returns tags, but the delegate one does. It adds a tag to distinguish the metrics from the immediate builder and the jittered one. So tags need to be in the interface.